geomstats/geometry/pullback_metric.py

```diff
@@ -103,6 +104,453 @@ def pickable_inner_product(i, j):
+
+    @abc.abstractmethod
+    def diffeomorphism(self, base_point):
+        r"""Diffeomorphism at base point.
+
+        Let :math:`f` be the diffeomorphism
+        :math:`f: M \rightarrow N` of the manifold
+        :math:`M` into the manifold `N`.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., *shape]
+            Base point.
+
+        Returns
+        -------
+        image_point : array-like, shape=[..., *i_shape]
+            Inner-product matrix.
+        """
+
+    def raw_jacobian_diffeomorphism(self, base_point):
+        r"""Raw jacobian of the diffeomorphism.
+
+        Raw jacobian autodiff of diffeomorphism regardless of vectorization.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., *shape]
+            Base point.
+
+        Returns
+        -------
+        mat : array-like, shape=[..., *i_shape, ..., *shape]
+            Inner-product matrix.
+        """
+        if self._raw_jacobian_diffeomorphism is None:
+            self._raw_jacobian_diffeomorphism = gs.autodiff.jacobian(
+                self.diffeomorphism
+            )
+        return self._raw_jacobian_diffeomorphism(base_point)
+
+    def jacobian_diffeomorphism(self, base_point):
+        r"""Jacobian of the diffeomorphism at base point.
+
+        Let :math:`f` be the diffeomorphism
+        :math:`f: M \rightarrow N` of the manifold
+        :math:`M` into the manifold `N`.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., *shape]
+            Base point.
+
+        Returns
+        -------
+        mat : array-like, shape=[..., *i_shape, *shape]
+            Inner-product matrix.
+        """
+        rad = base_point.shape[: -len(self.shape)]
+        base_point = gs.reshape(base_point, (-1,) + self.shape)
+
+        J = self.raw_jacobian_diffeomorphism(base_point)
+        J = gs.moveaxis(
+            gs.diagonal(J, axis1=0, axis2=len(self.embedding_metric.shape) + 1),
+            -1,
+            0,
+        )
+        J = gs.reshape(J, rad + self.embedding_metric.shape + self.shape)
+
+        return J
+
+    def tangent_diffeomorphism(self, tangent_vec, base_point):
+        r"""Tangent diffeomorphism at base point.
+
+        Let :math:`f` be the diffeomorphism
+        :math:`f: M \rightarrow N` of the manifold
+        :math:`M` into the manifold `N`.
+
+        df_p is a linear map from T_pM to T_f(p)N called
+        the tangent immesion
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., *shape]
+            Tangent vector at base point.
+
+        base_point : array-like, shape=[..., *shape]
+            Base point.
+
+        Returns
+        -------
+        image_tangent_vec : array-like, shape=[..., *i_shape]
+            Image tangent vector at image of the base point.
+        """
+        base_point, tangent_vec = gs.broadcast_arrays(base_point, tangent_vec)
```

Comment:
I think you are right it should be better to avoid mutual broadcast when a tangent vec is involved. `gs.broadcast_to(base_point, tangent_vec.shape)`. I will commit a change !